packages/test/test-end-to-end-tests/src/test/dataStoreRetrieval.spec.ts

```diff
@@ -0,0 +1,173 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { strict as assert } from "assert";
+import { IFluidHandle, IRequest } from "@fluidframework/core-interfaces";
+import { ITestObjectProvider } from "@fluidframework/test-utils";
+import {
+	describeFullCompat,
+	ITestDataObject,
+	TestDataObjectType,
+} from "@fluid-internal/test-version-utils";
+import { requestFluidObject } from "@fluidframework/runtime-utils";
+import { IContainerRuntimeBase } from "@fluidframework/runtime-definitions";
+
+/**
+ * These tests retrieve a data store via handle.get() during their initialization. They validate that
+ * retrieving a data store that was created locally works fine even if the outer data store has not
+ * finished initializing.
+ */
+describeFullCompat(
+	"data store retrieval during initialization tests",
+	(getTestObjectProvider, apis) => {
+		const {
+			dataRuntime: { DataObject, DataObjectFactory },
+			containerRuntime: { ContainerRuntimeFactoryWithDefaultDataStore },
+		} = apis;
+		class InnerDataObject extends DataObject implements ITestDataObject {
+			public get _root() {
+				return this.root;
+			}
+
+			public get _context() {
+				return this.context;
+			}
+
+			public get _runtime() {
+				return this.runtime;
+			}
+		}
+		const innerDataObjectFactory = new DataObjectFactory(
+			"InnerDataObject",
+			InnerDataObject,
+			[],
+			[],
+		);
+
+		class OuterDataObject extends DataObject implements ITestDataObject {
+			public get _root() {
+				return this.root;
+			}
+
+			public get _context() {
+				return this.context;
+			}
+
+			public get _runtime() {
+				return this.runtime;
+			}
+
+			private readonly innerDataStoreKey = "innerDataStore";
+
+			protected async initializingFirstTime(): Promise<void> {
+				const innerDataStore = await this._context.containerRuntime.createDataStore(
+					innerDataObjectFactory.type,
+				);
+				const innerDataObject = (await innerDataStore.entryPoint?.get()) as ITestDataObject;
+				this.root.set(this.innerDataStoreKey, innerDataObject.handle);
+				await innerDataObject.handle.get();
```

Comment:
```suggestion

				// Here is where we retrieve the inner object during outer object's initialization
				await innerDataObject.handle.get();
```

Reply:
Btw why did you put this here v. in `hasInitialized`?

Reply:
That was a mistake. It should be in `hasInitialized`